In [1]:
import requests
import pandas as pd
import matplotlib.pyplot as plt

%matplotlib inline

### For this project you will use `requests` package and the SODA API to access data through https://data.nashville.gov/. We'll start out with something familiar, the [Top 500 Monthly Searches](https://data.nashville.gov/Public-Services/Nashville-gov-Top-500-Monthly-Searches/fuaa-r5cm), then pull in different datasets further on. You will make different API requests for each individual question.

### Each dataset has its own api endpoint. You can find the endpoint for a dataset by clicking on the `API` button in the top right of the dataset screen, then copying the `API Endpoint`. The default output is `JSON`, which you can leave unchanged:

### ![api_endpoint](assets/api_endpoint.png)

### Each API is different, so it is very important to read the documentation for each API to know how to use it properly. The documentation for the SODA API is [here](https://dev.socrata.com/consumers/getting-started.html). It is **HIGHLY RECOMMENDED** that you read the documentation before making any requests, then do deeper dives into specific use cases when questions require. NOTE that the examples in the documentation don't use the `requests` package. You will need to look at the examples and figure out which things go in the `url` and which things go in the `params`.


## 1. Make an API request that returns the months where "fire" was searched in 2016. Which month had the most searches?  

In [23]:
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json'
params1 = {'year' : '2016', 'query_text' : 'fire' }

In [31]:
response1 = requests.get(endpoint, params1)

In [32]:
response1

<Response [200]>

In [45]:
result1 = response1.json()

In [46]:
result1

[{'month_name': 'January',
  'year': '2016',
  'query_count': '19',
  'query_text': 'fire'},
 {'month_name': 'February',
  'year': '2016',
  'query_count': '35',
  'query_text': 'fire'},
 {'month_name': 'March',
  'year': '2016',
  'query_count': '32',
  'query_text': 'fire'},
 {'month_name': 'April',
  'year': '2016',
  'query_count': '26',
  'query_text': 'fire'},
 {'month_name': 'May',
  'year': '2016',
  'query_count': '24',
  'query_text': 'fire'},
 {'month_name': 'June',
  'year': '2016',
  'query_count': '31',
  'query_text': 'fire'},
 {'month_name': 'July',
  'year': '2016',
  'query_count': '24',
  'query_text': 'fire'},
 {'month_name': 'August',
  'year': '2016',
  'query_count': '47',
  'query_text': 'fire'},
 {'month_name': 'September',
  'year': '2016',
  'query_count': '36',
  'query_text': 'fire'},
 {'month_name': 'October',
  'year': '2016',
  'query_count': '38',
  'query_text': 'fire'},
 {'month_name': 'November',
  'year': '2016',
  'query_count': '32',
  'query_text

In [47]:
result1[0]

{'month_name': 'January',
 'year': '2016',
 'query_count': '19',
 'query_text': 'fire'}

In [53]:
searches1_df = pd.DataFrame(result1)
searches1_df

month_name  year query_count query_text
0     January  2016          19       fire
1    February  2016          35       fire
2       March  2016          32       fire
3       April  2016          26       fire
4         May  2016          24       fire
5        June  2016          31       fire
6        July  2016          24       fire
7      August  2016          47       fire
8   September  2016          36       fire
9     October  2016          38       fire
10   November  2016          32       fire
11   December  2016          38       fire

In [54]:
searches1_df.sort_values(by = ['query_count'], ascending = False)

month_name  year query_count query_text
7      August  2016          47       fire
9     October  2016          38       fire
11   December  2016          38       fire
8   September  2016          36       fire
1    February  2016          35       fire
2       March  2016          32       fire
10   November  2016          32       fire
5        June  2016          31       fire
3       April  2016          26       fire
4         May  2016          24       fire
6        July  2016          24       fire
0     January  2016          19       fire

# August 2016 had the most searches of 'fire' with 47 instances.

## 2. Make an API request that returns all the times a query was run more than 100 times in a month. How many times did this occur?  

In [67]:
params2 = {"$where": "query_count>100", '$limit' : '25000'}
response2 = requests.get(endpoint, params2)
response2

<Response [200]>

In [68]:
result2 = response2.json()

In [69]:
result2

[{'month_name': 'March',
  'year': '2014',
  'query_count': '101',
  'query_text': 'permits'},
 {'month_name': 'January',
  'year': '2015',
  'query_count': '101',
  'query_text': 'criminal court clerk'},
 {'month_name': 'September',
  'year': '2015',
  'query_count': '101',
  'query_text': 'codes'},
 {'month_name': 'March',
  'year': '2016',
  'query_count': '101',
  'query_text': 'police'},
 {'month_name': 'March',
  'year': '2016',
  'query_count': '101',
  'query_text': 'civil service'},
 {'month_name': 'November',
  'year': '2016',
  'query_count': '101',
  'query_text': 'jobs'},
 {'month_name': 'November',
  'year': '2017',
  'query_count': '101',
  'query_text': 'metro holidays'},
 {'month_name': 'November',
  'year': '2017',
  'query_count': '101',
  'query_text': 'longevity pay'},
 {'month_name': 'January',
  'year': '2018',
  'query_count': '101',
  'query_text': 'West Nashville Heights Church of Christ'},
 {'month_name': 'January',
  'year': '2018',
  'query_count': '101',
 

In [70]:
searches2_df = pd.DataFrame(result2)
searches2_df

month_name  year query_count            query_text month
0         March  2014         101               permits   NaN
1       January  2015         101  criminal court clerk   NaN
2     September  2015         101                 codes   NaN
3         March  2016         101                police   NaN
4         March  2016         101         civil service   NaN
...         ...   ...         ...                   ...   ...
1257  September  2018         750     annual enrollment   NaN
1258    October  2018         816     annual enrollment   NaN
1259    January  2019        2646             Nashville     1
1260  September  2019        5327             directory     9
1261  September  2019        5610                  test     9

[1262 rows x 5 columns]

# There have been 1262 times a query occured more than 100 times

## 3. Make another API request that returns all the times "codes" was searched more than 100 times in a month. How many times did this occur?  

In [71]:
params3 = {'query_text' : 'codes', '$where': 'query_count>100', '$limit' : '25000'}
response3 = requests.get(endpoint, params3)
response3

<Response [200]>

In [73]:
result3 = response3.json()
result3

[{'month_name': 'September',
  'year': '2015',
  'query_count': '101',
  'query_text': 'codes'},
 {'month_name': 'April',
  'month': '4',
  'year': '2021',
  'query_count': '102',
  'query_text': 'codes'},
 {'month_name': 'August',
  'year': '2016',
  'query_count': '104',
  'query_text': 'codes'},
 {'month_name': 'November',
  'month': '11',
  'year': '2020',
  'query_count': '104',
  'query_text': 'codes'},
 {'month_name': 'December',
  'year': '2017',
  'query_count': '106',
  'query_text': 'codes'},
 {'month_name': 'May',
  'year': '2016',
  'query_count': '107',
  'query_text': 'codes'},
 {'month_name': 'April',
  'year': '2016',
  'query_count': '111',
  'query_text': 'codes'},
 {'month_name': 'July',
  'month': '7',
  'year': '2021',
  'query_count': '114',
  'query_text': 'codes'},
 {'month_name': 'January',
  'year': '2017',
  'query_count': '115',
  'query_text': 'codes'},
 {'month_name': 'May',
  'month': '5',
  'year': '2021',
  'query_count': '115',
  'query_text': 'codes'

In [75]:
searches3_df = pd.DataFrame(result3)
searches3_df

month_name  year query_count query_text month
0   September  2015         101      codes   NaN
1       April  2021         102      codes     4
2      August  2016         104      codes   NaN
3    November  2020         104      codes    11
4    December  2017         106      codes   NaN
5         May  2016         107      codes   NaN
6       April  2016         111      codes   NaN
7        July  2021         114      codes     7
8     January  2017         115      codes   NaN
9         May  2021         115      codes     5
10     August  2015         117      codes   NaN
11   February  2016         121      codes   NaN
12   February  2019         124      codes     2
13       June  2019         125      codes     6
14  September  2019         126      codes     9
15    January  2019         129      codes     1
16  September  2020         130      codes     9
17   February  2021         131      codes     2
18      March  2019         133      codes     3
19   November  2019         134      codes    11
20      March  2021         134      codes     3
21   November  2017         138      codes   NaN
22       June  2021         138      codes     6
23    October  2020         139      codes    10
24   December  2019         143      codes    12
25   February  2017         150      codes   NaN
26     August  2019         156      codes     8
27      March  2017         158      codes   NaN
28    October  2019         159      codes    10
29       July  2019         161      codes     7
30    January  2020         161      codes     1
31      March  2020         164      codes     3
32       July  2020         164      codes     7
33   December  2020         165      codes    12
34       July  2018         166      codes   NaN
35  September  2017         168      codes   NaN
36      April  2019         168      codes     4
37      April  2018         170      codes   NaN
38  September  2018         171      codes   NaN
39    October  2018         174      codes   NaN
40     August  2020         178      codes     8
41   February  2020         181      codes     2
42        May  2018         183      codes   NaN
43        May  2020         184      codes     5
44      March  2018         185      codes   NaN
45      April  2017         190      codes   NaN
46    January  2021         191      codes     1
47       June  2017         192      codes   NaN
48   November  2018         201      codes   NaN
49       June  2018         204      codes   NaN
50    October  2017         206      codes   NaN
51       June  2020         208      codes     6
52       July  2017         215      codes   NaN
53     August  2017         222      codes   NaN
54     August  2018         225      codes   NaN
55        May  2017         269      codes   NaN

# Codes has been searched over 100 times in a month 55 times.

## 4. Make an API request that returns the entire Top 500 Monthly Searches dataset. Make a chart that shows the number of times "maps" was searched in a month across the entire time frame.

In [82]:
params4 = {'$limit' : '50000'}
response4 = requests.get(endpoint, params4)
response4

<Response [200]>

In [83]:
results4 = response4.json()
results4

[{'month_name': 'January',
  'year': '2014',
  'query_count': '223',
  'query_text': 'ebid'},
 {'month_name': 'January',
  'year': '2014',
  'query_count': '112',
  'query_text': 'property maps'},
 {'month_name': 'January',
  'year': '2014',
  'query_count': '97',
  'query_text': 'maps'},
 {'month_name': 'January',
  'year': '2014',
  'query_count': '90',
  'query_text': 'property tax'},
 {'month_name': 'January',
  'year': '2014',
  'query_count': '76',
  'query_text': 'jobs'},
 {'month_name': 'January',
  'year': '2014',
  'query_count': '75',
  'query_text': 'employment'},
 {'month_name': 'January',
  'year': '2014',
  'query_count': '73',
  'query_text': 'human resources'},
 {'month_name': 'January',
  'year': '2014',
  'query_count': '62',
  'query_text': 'police'},
 {'month_name': 'January',
  'year': '2014',
  'query_count': '61',
  'query_text': 'criminal court clerk'},
 {'month_name': 'January',
  'year': '2014',
  'query_count': '53',
  'query_text': 'media releases'},
 {'mon

In [84]:
searches4_df = pd.DataFrame(results4)
searches4_df

month_name  year query_count         query_text month
0        January  2014         223               ebid   NaN
1        January  2014         112      property maps   NaN
2        January  2014          97               maps   NaN
3        January  2014          90       property tax   NaN
4        January  2014          76               jobs   NaN
...          ...   ...         ...                ...   ...
43672       July  2021           8     sign ordinance     7
43673       July  2021           8               snap     7
43674       July  2021           8     small business     7
43675       July  2021           8   water department     7
43676       July  2021           8  value of property     7

[43677 rows x 5 columns]

In [86]:
maps_searches = searches4_df.loc[searches4_df['query_text'] == 'maps']
# df.loc[df['column_name'] == some_value]
maps_searches.reset_index(level = 0, inplace = True)
maps_searches

index month_name  year query_count query_text month
0       2    January  2014          97       maps   NaN
1     503   February  2014         140       maps   NaN
2    1005      March  2014         144       maps   NaN
3    1503      April  2014         152       maps   NaN
4    2003        May  2014         119       maps   NaN
..    ...        ...   ...         ...        ...   ...
77  39196   November  2020          86       maps    11
78  39688   December  2020         113       maps    12
79  41188      March  2021         132       maps     3
80  41683      April  2021         148       maps     4
81  42682       June  2021         151       maps     6

[82 rows x 6 columns]

# Stretch Questions

## 5. Make an API request to pull back all the data from [hubNashville (311) Service Requests](https://data.nashville.gov/Public-Services/hubNashville-311-Service-Requests/7qhx-rexh) (check to see how many rows you can return in a single request). Compare it to the Top 500 Monthly Searches data set. What do you observe? (This is open-ended, there isn't a specific answer for this one)  

In [87]:
endpoint2 = 'https://data.nashville.gov/resource/7qhx-rexh.json'
response5 = requests.get(endpoint2)
results5 = response5.json()
stretch1_df = pd.DataFrame(results5)
stretch1_df

case_number       status                            case_request  \
0        151744  In Progress                     Property Violations   
1        198040       Closed  Resolved by hubNashville on First Call   
2        526941  In Progress              Streets, Roads & Sidewalks   
3         44230       Closed          Other Metro Services and Forms   
4        630756  In Progress                     Property Violations   
..          ...          ...                                     ...   
995      817801       Closed               Trash, Recycling & Litter   
996      819172       Closed                           Public Safety   
997      819005       Closed  Resolved by hubNashville on First Call   
998      818547       Closed  Resolved by hubNashville on First Call   
999      818196       Closed                           Public Safety   

                            case_subrequest  \
0                     Residential Complaint   
1    Resolved by hubNashville on First Call   
2                                  Potholes   
3                                     Other   
4                          Fence Complaints   
..                                      ...   
995    Missed Pickup - Recycle Cart Service   
996       Control Number Request for Towing   
997  Resolved by hubNashville on First Call   
998  Resolved by hubNashville on First Call   
999       Control Number Request for Towing   

                      additional_subrequest         date_time_opened  \
0                     Residential Complaint  2019-03-08T15:12:07.000   
1    Resolved by hubNashville on First Call  2019-06-11T15:46:08.000   
2                                     Alley  2020-12-18T18:43:02.000   
3                                     Other  2018-02-06T15:21:35.000   
4                          Fence Complaints  2021-03-21T21:48:00.000   
..                                      ...                      ...   
995  Recycle - Curbside/Alley Missed Pickup  2021-11-03T19:27:16.000   
996       Control Number Request for Towing  2021-11-05T00:06:06.000   
997  Resolved by hubNashville on First Call  2021-11-04T20:16:57.000   
998  Resolved by hubNashville on First Call  2021-11-04T15:33:17.000   
999       Control Number Request for Towing  2021-11-04T12:41:08.000   

                case_origin state_issue closed_when_created  \
0    hubNashville Community       False               False   
1                     Phone       False                True   
2                     Phone       False               False   
3                     Phone       False               False   
4    hubNashville Community       False               False   
..                      ...         ...                 ...   
995  hubNashville Community       False               False   
996              Mobile App       False                True   
997                   Phone       False                True   
998                   Phone       False                True   
999              Mobile App       False                True   

         incident_address  ... :@computed_region_v67z_xm3t  \
0    4000 Anderson Rd #76  ...                           9   
1                     NaN  ...                         NaN   
2          107 28th Ave N  ...                           6   
3                     NaN  ...                         NaN   
4           402 N 16th St  ...                           7   
..                    ...  ...                         ...   
995         3304 Hobbs Rd  ...                           6   
996   801 Park Terrace Dr  ...                           9   
997                   NaN  ...                         NaN   
998                   NaN  ...                         NaN   
999                  I-65  ...                         NaN   

    :@computed_region_kh5x_g7w5 :@computed_region_cfa7_hbpz  \
0                            27                         731   
1                           NaN                         NaN   
2                 

In [ ]:
params5

## 6. Find 2 new data sets on data.nashville.gov, make API requests to pull the data, and do an analysis that combines the data sets.            

# Bonus

## 7. Socrata is used by many cities, states, and federal organizations. Find additional datasets through [Socrata's Open Data Network](http://www.opendatanetwork.com/) and do an analysis comparing them to Nashville or each other.

# Show and Tell

## At the end of the project you will present some general insights, visualizations, or other finding from any part of the project. This will be informal (showing your Jupyter notebook is fine, no need to make a powerpoint) and should be no more than 5 min. If you had challenges making your visualizations, then it is fine to discuss your experience working with the API and what you were intending to show.
